File: TickTalk/app/src/main/java/com/example/adeeliftikhar/ticktalk/ProfileActivity.java
package com.example.adeeliftikhar.ticktalk;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;
import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import id.zelory.compressor.Compressor;

public class ProfileActivity extends AppCompatActivity {
    Button buttonChangeProfilePhoto, buttonUpdateProfileStatus;
    TextView textViewUserName, textViewUserStatus;
    ImageView imageViewProfile;

    FirebaseAuth mAuth;
    FirebaseUser currentUser;
    String currentUserId;
    DatabaseReference dbReference;
    FirebaseApp firebaseApp;
    StorageReference storageRef;
    int galleryPic = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_profile);
        imageViewProfile = findViewById(R.id.image_view_profile);
        textViewUserName = findViewById(R.id.txt_view_user_name);
        textViewUserStatus = findViewById(R.id.txt_view_user_status);
        buttonChangeProfilePhoto = findViewById(R.id.button_change_profile_photo);
        buttonUpdateProfileStatus = findViewById(R.id.button_update_profile_status);
//        Initializing the Database...
        firebaseApp = FirebaseApp.initializeApp(this);
//        Getting Instance(Object) of Firebase Application (e.g ==> TickTalk)...
        mAuth = FirebaseAuth.getInstance();
//        Getting User From Firebase....
        currentUser = mAuth.getCurrentUser();
//        Getting User ID from current user....
        if (currentUser != null) {
            currentUserId = currentUser.getUid();
        }
//        Getting Current User...
        dbReference = FirebaseDatabase.getInstance().getReference().child("Users").child(currentUserId);
//        App will be run offline...
        dbReference.keepSynced(true);
        RetrieveDataOfTheUser();
        storageRef = FirebaseStorage.getInstance().getReference();
    }

    private void RetrieveDataOfTheUser() {
//        Used to Retrieve Value from Database...
        dbReference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                String name = dataSnapshot.child("name").getValue().toString();
                String status = dataSnapshot.child("status").getValue().toString();
                final String image = dataSnapshot.child("image").getValue().toString();
                String thumb_image = dataSnapshot.child("thumb_image").getValue().toString();
                textViewUserName.setText(name);
                textViewUserStatus.setText(status);
                if (!image.equals("Default")) {
                    Picasso.get().load(image).placeholder(R.drawable.bg).into(imageViewProfile, new Callback() {
                        //                       Callback ==> Response
                        @Override
                        public void onSuccess() {
//                            Toast.makeText(ProfileActivity.this, "Updated Successfully", Toast.LENGTH_SHORT).show();
                        }

                        @Override
                        public void onError(Exception e) {
                            Picasso.get().load(image).placeholder(R.drawable.ic_launcher_foreground).into(imageViewProfile);
                        }
                    });
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });
    }

    public void changePhoto(View view) {
//        CropImage ==> It is a library which is used to get image from external apps (e.g gallery or camera etc)
//        by executing intent automatically (e.g startActivityForResult).
        CropImage.activity().setGuidelines(CropImageView.Guidelines.ON).start(ProfileActivity.this);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == galleryPic && resultCode == RESULT_OK) {
//            Now coming picture is stored in image.
            Uri image = data.getData();
//            Will crop the image.
            CropImage.activity(image).setAspectRatio(1, 1).start(ProfileActivity.this);
        }
        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
//            Now picture is stored in result.
            CropImage.ActivityResult result = CropImage.getActivityResult(data);
            if (resultCode == RESULT_OK) {
                final ProgressDialog progressDialog = new ProgressDialog(ProfileActivity.this);
                progressDialog.setTitle("OK");
                progressDialog.setMessage("Image Uploading... ");
                progressDialog.setCancelable(false);
                progressDialog.show();
//                resultUri has the current Image which is not compress yet(will store in image_filepath).
                Uri resultUri = result.getUri();
//                Now Picture is in fileImage and sending to the compression method (compressImage).
                File fileImage = new File(resultUri.getPath());
//                byteImageCompress has the Compressed image (will store in thumb_filepath).
                final byte[] byteImageCompress = compressImage(fileImage);
//                Storage path of image and thumb image.
                StorageReference image_filepath = storageRef.child("profile_images").child(currentUserId + ".jpg");
                final StorageReference thumb_filepath = storageRef.child("profile_images").child("thumbs").child(currentUserId + ".jpg");
//                Now putting image by using that path.
                image_filepath.putFile(resultUri).addOnCompleteListener(new OnCompleteListener<UploadTask.TaskSnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<UploadTask.TaskSnapshot> task) {
                        if (task.isSuccessful()) {
                            final String imageDownloadLink = task.getResult().getDownloadUrl().toString();
                            UploadTask uploadTask = thumb_filepath.putBytes(byteImageCompress);
                            uploadTask.addOnCompleteListener(new OnCompleteListener<UploadTask.TaskSnapshot>() {
                                @Override
                                public void onComplete(@NonNull Task<UploadTask.TaskSnapshot> task) {
                                    String thumbNailDownloadLink = task.getResult().getDownloadUrl().toString();
                                    if (task.isSuccessful()) {
                                        Map updateHashMap = new HashMap();
                                        updateHashMap.put("image", imageDownloadLink);
                                        updateHashMap.put("thumb_image", thumbNailDownloadLink);
                                        dbReference.updateChildren(updateHashMap).addOnCompleteListener(new OnCompleteListener() {
                                            @Override
                                            public void onComplete(@NonNull Task task) {
                                                progressDialog.dismiss();
//                                                Toast.makeText(ProfileActivity.this, "Successfully Updated the Profile", Toast.LENGTH_SHORT).show();
                                            }
                                        });
                                    }
                                }
                            });
                        }
                    }
                });
            }
        }
    }

    private byte[] compressImage(File imageFile) {
        Bitmap imageBitmap;
        byte[] thumbByte = new byte[0];
        try {
            imageBitmap = new Compressor(this).
                    setMaxWidth(200)
                    .setMaxHeight(200)
                    .setQuality(75)
                    .compressToBitmap(imageFile);
//      Creating byte Stream
//            Input/output Stream controls the input and output of data.
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            imageBitmap.compress(Bitmap.CompressFormat.JPEG, 100, baos);
//            Now outputStream data send and retrieve data in the form of byteArray.
            thumbByte = baos.toByteArray();
            return thumbByte;

        } catch (IOException e) {
            e.printStackTrace();
        }
        return thumbByte;
    }

    public void updateStatus(View view) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(ProfileActivity.this);
        builder.setTitle("Update Your Status");
        builder.setMessage("Enter New Status");
        final EditText editText = new EditText(this);
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
        );
        editText.setLayoutParams(layoutParams);
        editText.setHint("Update Status");
        builder.setView(editText);
        builder.setCancelable(false);

        builder.setPositiveButton("Update", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                String status = editText.getText().toString();
                dbReference.child("status").setValue(status).addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        if (task.isSuccessful()) {
//                            Toast.makeText(ProfileActivity.this, "Status Updated Successfully", Toast.LENGTH_SHORT).show();
                        }
                    }
                });
            }
        });

        builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        });
        builder.show();
    }
}

File: Messenger/app/src/main/java/com/example/adeeliftikhar/messenger/ProfileActivity.kt
package com.example.adeeliftikhar.messenger

import android.app.Activity
import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import com.example.adeeliftikhar.messenger.Internet.CheckInternetConnectivity
import com.example.adeeliftikhar.messenger.Models.RegisterModel
import com.example.adeeliftikhar.messenger.SessionsManager.LoginSessionManager
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.EmailAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import com.theartofdev.edmodo.cropper.CropImage
import com.theartofdev.edmodo.cropper.CropImageView
import de.hdodenhof.circleimageview.CircleImageView
import id.zelory.compressor.Compressor
import kotlinx.android.synthetic.main.activity_profile.*
import kotlinx.android.synthetic.main.activity_register.*
import kotlinx.android.synthetic.main.profile_alert_dialog.view.*
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException

class ProfileActivity : AppCompatActivity() {
    var currentUser: FirebaseUser? = null
    var currentUserId = ""
    var dbRef: DatabaseReference? = null
    var storageRef: StorageReference? = null
    private var downloadImageUri: String? = null
    private val galleryPic = 1
    var gotImage: Boolean = false
    var count: Int = 0

    var id = ""
    var name = ""
    var email = ""
    var password = ""
    var image = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_profile)

        supportActionBar?.title = "Profile"
        checkUser()
        retrieveUserProfileData()

        button_update_user_profile.setOnClickListener {
            showUpdateAlertDialog(name, email, image)
        }

        button_delete_user_profile.setOnClickListener {
            deleteAlertBox()
        }
    }

    companion object {
        var imageViewUser: CircleImageView? = null
    }

    private fun checkUser() {
        val mAuth = FirebaseAuth.getInstance()
//        Getting User From Firebase....
        currentUser = mAuth.getCurrentUser()
//        Getting User ID from current user....
        if (currentUser != null) {
            currentUserId = currentUser!!.getUid()
        }
        dbRef = FirebaseDatabase.getInstance().reference.child("UsersInfo").child(currentUserId)
        storageRef =
                FirebaseStorage.getInstance().getReference().child("Images").child("UsersImages").child(currentUserId)
        dbRef?.keepSynced(true)
    }

    private fun retrieveUserProfileData() {
        dbRef?.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (count != 1) {
                    id = dataSnapshot.child("id").value!!.toString()
                    name = dataSnapshot.child("name").value!!.toString()
                    email = dataSnapshot.child("email").value!!.toString()
                    password = dataSnapshot.child("password").value!!.toString()
                    image = dataSnapshot.child("image").value!!.toString()
                    text_view_user_profile_name.setText(name)
                    text_view_user_profile_email.setText(email)
                    if (image != "Default") {
                        Picasso.get().load(image).placeholder(R.drawable.common_pic_place_holder)
                            .into(image_view_profile, object : Callback {
                                override fun onSuccess() {
//                            Toast.makeText(ProfileActivity.this, "Updated Successfully", Toast.LENGTH_SHORT).show();
                                }

                                override fun onError(e: Exception) {
                                    Picasso.get().load(image).placeholder(R.drawable.common_pic_place_holder)
                                        .into(image_view_profile)
                                }
                            })
                    }
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })
    }

    private fun changePhoto() {
        CropImage.activity().setGuidelines(CropImageView.Guidelines.ON).start(this@ProfileActivity)
    }

    private fun showUpdateAlertDialog(name: String, email: String, image: String) {
        val builder = AlertDialog.Builder(this@ProfileActivity)
        val view = LayoutInflater.from(this@ProfileActivity).inflate(R.layout.profile_alert_dialog, null)

        imageViewUser = view.image_view_profile_alert
        val imageUri = image
        if (image != "image") {
            Picasso.get().load(image).placeholder(R.drawable.common_pic_place_holder)
                .into(imageViewUser, object : Callback {
                    override fun onSuccess() {
                    }

                    override fun onError(e: Exception) {
                        Picasso.get().load(image).placeholder(R.drawable.common_pic_place_holder)
                            .into(imageViewUser)
                    }
                })
        }

        imageViewUser?.setOnClickListener { changePhoto() }

        view.edit_text_user_profile_name_alert.setText(name)
        view.edit_text_user_profile_email_alert.setText(email)

        view.button_update_user_profile_alert.setOnClickListener {
            updateUserEmailAndProfile(view, imageUri)
        }

        builder.setView(view)
        builder.setCancelable(false)

        val alertDialog = builder.show()
        view.button_dismiss.setOnClickListener {
            alertDialog.dismiss()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == galleryPic && resultCode == Activity.RESULT_OK) {
            val image = data!!.data
            CropImage.activity(image).setAspectRatio(1, 1).start(this@ProfileActivity)
        }
        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            val result = CropImage.getActivityResult(data)
            if (resultCode == Activity.RESULT_OK) {

                val resultUri = result.getUri()
                val fileImage = File(resultUri.getPath())
                val compressImage = compressImage(fileImage)

                val bitmapImage = BitmapFactory.decodeByteArray(compressImage, 0, compressImage.size)
                imageViewUser?.setImageBitmap(bitmapImage)
                gotImage = true

                if (resultUri == null) return
//                val fileName = UUID.randomUUID().toString()
                storageRef?.putFile(resultUri)!!.addOnSuccessListener {
                    storageRef!!.downloadUrl.addOnSuccessListener {
                        downloadImageUri = it.toString()
                    }
                }
            }
        }
    }

    private fun compressImage(imageFile: File): ByteArray {
        val imageBitmap: Bitmap
        var thumbByte = ByteArray(0)
        try {
            imageBitmap = Compressor(this@ProfileActivity).setMaxWidth(200)
                .setMaxHeight(200)
                .setQuality(75)
                .compressToBitmap(imageFile)
            val baos = ByteArrayOutputStream()
            imageBitmap.compress(Bitmap.CompressFormat.JPEG, 100, baos)
            thumbByte = baos.toByteArray()
            return thumbByte

        } catch (e: IOException) {
            e.printStackTrace()
        }

        return thumbByte
    }

    private fun updateUserEmailAndProfile(view: View, imageUri: String) {
        count = 1

        val currentUser = FirebaseAuth.getInstance().currentUser?.uid.toString()
        val id = currentUser
        val dbRef = FirebaseDatabase.getInstance().getReference("UsersInfo/$currentUser")

        if (downloadImageUri.isNullOrEmpty()) {
            downloadImageUri = imageUri
            gotImage = true
        }

        if (name.isEmpty() || email.isEmpty() || password.isEmpty() || !gotImage) {
            showSnackBar()
        } else {
//            Checking Internet Connectivity...
            if (!CheckInternetConnectivity.isConnected(this)) {
                Toast.makeText(this, "No Internet Connection", Toast.LENGTH_SHORT).show()
            } else {

                Toast.makeText(this, "Updating Profile, Plz wait...", Toast.LENGTH_SHORT).show()
                name = view.edit_text_user_profile_name_alert.text.toString()
                email = view.edit_text_user_profile_email_alert.text.toString()

                val user = FirebaseAuth.getInstance().currentUser ?: return
//        Getting values from session manager...
                val credential = EmailAuthProvider.getCredential(
                    LoginSessionManager(this).getUserEmail(),
                    LoginSessionManager(this).getUserPassword()
                )
                user.reauthenticate(credential).addOnCompleteListener {
                    Log.d("Update", "User re-authenticated.")
                    user.updateEmail(email).addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            Log.d("Update", "User email address updated.")

//                        Update Profile Now...
                            val registerModel = RegisterModel(id, name, email, password, downloadImageUri.toString())

                            dbRef.setValue(registerModel).addOnCompleteListener {
                                if (it.isSuccessful) {
                                    Toast.makeText(this, "Profile Updated Successfully", Toast.LENGTH_SHORT).show()

//                        Update Login Session
                                    val loginSessionManager = LoginSessionManager(this@ProfileActivity)
                                    loginSessionManager.loginTheUser(true, email, password)

                                    val intent = Intent(this, MainActivity::class.java)
                                    startActivity(intent)
                                    finish()
                                }
                            }
                                .addOnFailureListener() {
                                    Log.d("Updated", "${it.message}")
                                    spin_kit_view_register.visibility = View.GONE
                                    Toast.makeText(
                                        this,
                                        "Failed to Update Data:  ${it.message}",
                                        Toast.LENGTH_SHORT
                                    ).show()
                                }
                        }
                    }
                }
            }
        }
    }

    private fun showSnackBar() {
        val linearLayout = linear_layout_profile
        val snackbar =
            Snackbar.make(linearLayout, "Please Fill All Fields \n And Insert an Image", Snackbar.LENGTH_SHORT)
                .setActionTextColor(Color.WHITE).setAction("Ok") { }
        val snackBarView = snackbar.view
        snackBarView.setBackgroundColor(Color.parseColor("#BF360C"))
        snackbar.show()
    }

    private fun deleteAlertBox() {
        val builder = android.support.v7.app.AlertDialog.Builder(this@ProfileActivity)
        builder.setTitle("Delete")
        builder.setMessage("Do you want to Delete your Account?")
        builder.setPositiveButton("Delete") { dialog, which ->
            if (!CheckInternetConnectivity.isConnected(this@ProfileActivity)) {
                Toast.makeText(this@ProfileActivity, "No Internet Connection", Toast.LENGTH_SHORT).show()
            } else {
                deleteProgressDialog()
                deleteAccount()
            }
        }.setNegativeButton("Cancel") { dialog, which -> }
        builder.show()
    }

    private fun deleteProgressDialog() {
        val progressDialog = ProgressDialog(this)
        progressDialog.setMessage("Deleting Your Account.")
        progressDialog.setCancelable(false)
        progressDialog.show()
    }

    private fun deleteAccount() {
        count = 1
        deleteFromFirebase()
    }

    private fun deleteFromFirebase() {
        val user = FirebaseAuth.getInstance().currentUser ?: return

//        Getting values from session manager...
        val delEmail = LoginSessionManager(this).getUserEmail()
        val delPassword = LoginSessionManager(this).getUserPassword()

        val credential = EmailAuthProvider.getCredential(delEmail, delPassword)

        user.reauthenticate(credential).addOnCompleteListener {
            user.delete().addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    storageRef!!.delete().addOnCompleteListener {
                        if (task.isSuccessful) {
                            dbRef?.removeValue()?.addOnCompleteListener {
                                if (it.isSuccessful) {
                                    Toast.makeText(this, "Your Profile is Deleted", Toast.LENGTH_SHORT).show()
                                    clearLoginSession()
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private fun clearLoginSession() {
        val loginSessionManager = LoginSessionManager(this@ProfileActivity)
        loginSessionManager.loginTheUser(false, "", "")
        FirebaseAuth.getInstance().signOut()
        val intent = Intent(this, RegisterActivity::class.java)
        startActivity(intent)
        finish()
    }
}
